core/store/src/trie/split.rs

```diff
@@ -738,8 +837,228 @@ mod tests {
+        use near_primitives::types::{AccountId, Balance, StateChangeCause};
+        use rand::RngCore;
+        use rand::distributions::{Distribution, Uniform};
+        use rand::prelude::SliceRandom;
+        use rand::rngs::StdRng;
+        use rand::{Rng, SeedableRng};
+        use std::fmt::Debug;
+
+        impl<NodePtr, Value, Storage> TrieDescent<NodePtr, Value, Storage>
+        where
+            NodePtr: Debug + Copy,
+            Storage: GenericTrieInternalStorage<NodePtr, Value>,
+        {
+            /// Helper method to check the split yielded by a pre-defined key.
+            fn check_split(mut self, key_bytes: &[u8]) -> TrieSplit {
+                let key_nibbles = key_bytes
+                    .iter()
+                    .flat_map(|byte| {
+                        let (nib1, nib2) = byte_to_nibbles(*byte);
+                        [nib1, nib2]
+                    })
+                    .collect_vec();
+
+                for nibble in key_nibbles {
+                    let children_mem_usage = self.aggregate_children_mem_usage();
+                    let child_mem_usage = children_mem_usage[nibble as usize];
+                    let left_mem_usage = children_mem_usage[..nibble as usize].iter().sum();
+                    self.descend_step(nibble, child_mem_usage, left_mem_usage);
+                }
+
+                TrieSplit::new(
+                    self.nibbles,
+                    self.left_memory,
+                    self.right_memory + self.middle_memory,
+                )
+            }
+        }
+
+        /// Helper function to get the split yielded by the given boundary account.
+        /// Assumes the trie is using memtries.
+        fn get_memtrie_split(trie: &Trie, boundary_account: &AccountId) -> TrieSplit {
+            let key_bytes = boundary_account.as_bytes();
+            let memtries = trie.lock_memtries().unwrap();
+            let trie_storage = MemTrieIteratorInner::new(&memtries, trie);
+            TrieDescent::new(trie_storage).check_split(key_bytes)
+        }
+
+        impl TrieSplit {
+            /// Get difference between right and left memory
+            fn mem_diff(&self) -> i64 {
+                self.right_memory as i64 - self.left_memory as i64
+            }
+        }
+
+        #[test]
+        fn big_random_trie() {
+            // Seeds between 0 and 10000 were tested and these were picked as they uncovered
+            // some bugs and corner cases.
+            big_random_trie_inner(1).unwrap();
+            big_random_trie_inner(4).unwrap();
+            big_random_trie_inner(11).unwrap();
+            big_random_trie_inner(369).unwrap();
+        }
+
+        fn big_random_trie_inner(seed: u64) -> anyhow::Result<()> {
+            // Initialize an empty trie and start an update to populate it with some data
+            let (shard_tries, layout) =
+                TestTriesBuilder::new().with_in_memory_tries(true).with_flat_storage(true).build2();
```

Comment:
If we already support partial storage, can we also test that `find_trie_split` called on recorded storage returns the same result? The purpose of this check is that proof is sufficient to end up with the same trie split. There is bunch of other tests with the same builder setup.